boot/freeldr/freeldr/ntldr/arch/i386/winldr.c

```diff
@@ -270,64 +338,143 @@ MempAllocatePTE(ULONG Entry, PHARDWARE_PTE *PhysicalPT, PHARDWARE_PTE *KernelPT)
+
+    PdIdx = (PdeIdx / PAE_PDE_PER_PAGE);
+
+    Pde = &PaeTables->PaePd[PdIdx].PaePde[PdeIdx & PAE_PDE_MASK];
+    Pde->PageFrameNumber = (ULONG)*PhysicalPde >> MM_PAGE_SHIFT;
+    Pde->Valid = 1;
+    Pde->Write = 1;
+
+
+    if ((PdeIdx + KERNEL_PDE_IDX) > (MAX_PAE_PDE_COUNT - 1))
+    {
+        ERR("WARNING! PdeIdx %X > %X\n", (PdeIdx + KERNEL_PDE_IDX, (MAX_PAE_PDE_COUNT - 1)));
+    }
+
+    /* Kernel-mode mapping */
+    *KernelPde = (PHARDWARE_PDE_X86_PAE)&KernelPageTablesBuffer[KernelPageTables * MM_PAGE_SIZE];
+    KernelPageTables++;
+
+    PdIdx = ((PdeIdx + KERNEL_PDE_IDX) / PAE_PDE_PER_PAGE);
+
+    Pde = &PaeTables->PaePd[PdIdx].PaePde[PdeIdx & PAE_PDE_MASK];
+    Pde->PageFrameNumber = (ULONG)*PhysicalPde >> MM_PAGE_SHIFT;
+    Pde->Valid = 1;
+    Pde->Write = 1;
+}
+
 BOOLEAN
 MempSetupPaging(IN PFN_NUMBER StartPage,
                 IN PFN_COUNT NumberOfPages,
                 IN BOOLEAN KernelMapping)
 {
-    PHARDWARE_PTE PhysicalPT;
-    PHARDWARE_PTE KernelPT;
-    PFN_COUNT Entry, Page;
-
-    TRACE("MempSetupPaging: SP 0x%X, Number: 0x%X, Kernel: %s\n",
-       StartPage, NumberOfPages, KernelMapping ? "yes" : "no");
+    PFN_COUNT Page;
+    PFN_COUNT PdeIdx;
+    PFN_COUNT PhysicalPdIdx;
+    PFN_COUNT KernelPdIdx;
 
-    // HACK
-    if (StartPage+NumberOfPages >= 0x80000)
+    /* HACK */
+    if ((StartPage + NumberOfPages) >= 0x80000)
     {
-        //
-        // We cannot map this as it requires more than 1 PDE
-        // and in fact it's not possible at all ;)
-        //
-        //Print(L"skipping...\n");
+        /* We cannot map this as it requires more than 1 PDE
+           and in fact it's not possible at all ;)
+        */
+        ERR("MempSetupPaging: skipping... StartPage %X, Pages %X, Kernel: %s\n",
+            StartPage, NumberOfPages, KernelMapping ? "yes" : "no");
+
         return TRUE;
     }
 
-    //
-    // Now actually set up the page tables for identity mapping
-    //
-    for (Page = StartPage; Page < StartPage + NumberOfPages; Page++)
-    {
-        Entry = Page >> 10;
+    TRACE("MempSetupPaging: StartPage %X, Pages %X, Kernel: %s\n", StartPage, NumberOfPages, KernelMapping ? "yes" : "no");
 
-        if (((PULONG)PDE)[Entry] == 0)
+    /* Now actually set up the page tables for identity mapping */
+    for (Page = StartPage; Page < (StartPage + NumberOfPages); Page++)
+    {
+        if (!PaeModeOn)
         {
-            MempAllocatePTE(Entry, &PhysicalPT, &KernelPT);
+            PHARDWARE_PDE_X86 PhysicalPde;
+            PHARDWARE_PDE_X86 KernelPde;
+
+            PdeIdx = (Page >> 10);
+
+            if (((PULONG)PDE)[PdeIdx] == 0)
+            {
+                MempAllocatePde(PdeIdx, &PhysicalPde, &KernelPde);
+            }
+            else
+            {
+                PhysicalPde = (PHARDWARE_PTE)(PDE[PdeIdx].PageFrameNumber << MM_PAGE_SHIFT);
+                KernelPde = (PHARDWARE_PTE)(PDE[PdeIdx+(KSEG0_BASE >> 22)].PageFrameNumber << MM_PAGE_SHIFT);
```

Comment:
22 == PDE_SHIFT?
Also would be nice to define `KERNEL_PDE_IDX` for non-PAE config as well

Reply:
Yes, PDE_SHIFT, for non-PAE it 10-10-12 (page directory - page table - page offset).
I don't want to touch non-PAE code right now...